### ISSUE TITLE
Strange behaviour with Resize

### ISSUE THREAD

**Issue by greut** [h2non/bimg#46]
Hi,

I'm trying to make sense of the `Resize` function. From the `resize_test.go` file, I've adapted the following.

* `TestResize` doesn't affect the source file and the output is a copy of it
* `TestResize2` (bad naming) does work and produces the smaller file.

I don't get why in `affineTransform` (`resize.go:256`), the tranformation just made is rejected.

```go
func TestResize(t *testing.T) {
	options := Options{Width: 800, Height: 600}
	buf, _ := Read("fixtures/test.jpg")

	newImg, err := Resize(buf, options)
	if err != nil {
		t.Errorf("Resize(imgData, %#v) error: %#v", options, err)
	}

	size, _ := NewImage(newImg).Size()
	if size.Width > options.Width || size.Height > options.Height {
		t.Errorf("New image bigger than the options. %#v", size)
	}
}

func TestResize2(t *testing.T) {
	options := Options{Width: 105, Height: 66}
	buf, _ := Read("fixtures/test.jpg")

	newImg, err := Resize(buf, options)
	if err != nil {
		t.Errorf("Resize(imgData, %#v) error: %#v", options, err)
	}

	size, _ := NewImage(newImg).Size()
	if size.Width > options.Width || size.Height > options.Height {
		t.Errorf("New image bigger than the options. %#v", size)
	}
}
```

Cheers!

**Comment by h2non** [comment]
Have you tried passing `Enlage: true` option?

**Comment by greut** [comment]
Thanks. With `Crop: true`, the output image is the one given but it wasn't what I wanted to do in the first place. The change you've made does a crop in my case.

The behaviour of `Resize` is still not very clear to me.

Let `i` be an image of `1000 x 1000`. This is what I would expect, probably.
* `Resize(i, {Width: 800, Height: 600})` produces a non-cropped, deformed `800 x 600` picture.
* `Resize(i, {Width: 800, Height: 600, Crop: true})` produces a cropped, non-deformed `800x600` picture
* `Resize(i, {Width: 800})` produces a non-cropped, non-deformed `800 x 800` picture
* `Resize(i, {Height: 800})` ditto
* `Resize(i, {Width: 2000, Height: 1500})` produces the original picture (_this is the very inconsistent one_)
* `Resize(i, {Width: 2000, Height: 1500, Enlarge: true})` produces a non-cropped, deformed `2000 x 1500` picture
* `Resize(i, {Width: 2000, Height: 1500, Crop: true, Enlarge: true})` produces a cropped, non-deformed `2000 x 1500` picture (`2000 x 2000` before the crop)
* `Resize(i, {Width: 2000, Enlarge: true})` produces a `2000 x 2000` picture.

**Comment by h2non** [comment]
I see. My changes were just experimental. I totally agree with your expectations. I'll dig into it in detail

**Comment by greut** [comment]
k, thanks!

**Comment by h2non** [comment]
@greut This should be fixed now. A couple of notes about this:

- The resize behavior was, indeed, wrong (therefore this was considered as `bug`)
- All the expectations you described above should work now, expect for one case:
  - If only one of `Width` or `Height` options are passed, the image ratio will be maintained. I think that's the most coherent and expected behavior. E.g: given an image of `2000x1000`, if only passes `Width: 1000`, the resultant size will be: `1000x500`. In case that you want a custom fixed size, you must pass both `Width` and `Height`. I believe that explicit is better than implicit. And more simple.
- I've created [a test suite](https://github.com/h2non/bimg/blob/master/resize_test.go#L31) to cover a bunch of cases. 
- A new option was added, called `Force: bool`. This rarely will be used from API consumers, but in cases like yours could be useful to explicitly force a resize operation.
- A new shortcut method called `ForceResize(width, height)` was added in `Image` struct. A bit of DSL sometimes is handy.

I've just released `0.1.18` version, so feel free to update your package!
